scripts/lib/helpers.ts

```diff
@@ -1,63 +1,228 @@
 import { existsSync, readFileSync, writeFileSync } from "fs";
+import { relative, resolve } from "path";
 
 import admin from "firebase-admin";
 import formatDate from "date-fns/format/index.js";
-import { resolve } from "path";
 
-export const makeSaveToBackupFile = (filenamePrefix: string) => (
-  data: string | {} | [],
-  type: string = "general",
-  ext: string = "backup.json"
-) => {
-  const backupDate = formatDate(new Date(), "yyyyMMdd-HHmmss");
-  const backupFilename = `${backupDate}-${filenamePrefix}-${type}.${ext}`;
-
-  const jsonData = JSON.stringify(data, null, 2);
+import { Table } from "../../src/types/Table";
 
-  writeFileSync(backupFilename, jsonData, "utf8");
-
-  console.log(`Saved backup (type='${type}') to ${backupFilename}`);
-};
+export interface CredentialFile {
+  type?: string;
+  project_id?: string;
+  private_key_id?: string;
+  private_key?: string;
+  client_email?: string;
+  client_id?: string;
+  auth_uri?: string;
+  token_uri?: string;
+  auth_provider_x509_cert_url?: string;
+  client_x509_cert_url?: string;
+}
 
+/**
+ * Initialise a new Firebase Admin App
+ *
+ * @param projectId Firebase project ID
+ * @param appName An arbitrary name to reference this app by within the script (mostly
+ *   useful when connecting to 2 projects in the same script, eg. to copy data between them)
+ * @param credentialPath Path to the Firebase credential file to be used
+ *
+ * @example Basic Usage
+ *   initFirebaseAdminApp(projectId);
+ *
+ * @example Resolving/using a specific credential file (if defined)
+ *   initFirebaseAdminApp(projectId, {
+ *     credentialPath: credentialPath
+ *       ? resolve(__dirname, credentialPath)
+ *       : undefined,
+ *   });
+ *
+ * @example Initialising two projects at once (eg. to copy data between them)
+ *   const sourceApp = initFirebaseAdminApp(SOURCE_PROJECT_ID, {
+ *     appName: "sourceApp",
+ *     credentialPath: resolve(__dirname, SOURCE_CREDENTIAL_FILE),
+ *   });
+ *
+ *   const destApp = initFirebaseAdminApp(DEST_PROJECT_ID, {
+ *     appName: "destApp",
+ *     credentialPath: resolve(__dirname, DEST_CREDENTIAL_FILE),
+ *   });
+ */
 export const initFirebaseAdminApp = (
   projectId: string,
   {
     appName,
-    credentialPath = resolve(__dirname, "../prodAccountKey.json"),
-  }: { appName?: string; credentialPath?: string }
-): admin.app.App =>
-  admin.initializeApp(
+    credentialPath: _credentialPath,
+  }: { appName?: string; credentialPath?: string } = {}
+): admin.app.App => {
+  const credentialPath = _credentialPath ?? resolveDefaultCredentialPath();
```

Comment:
The documentation accepts `serviceAccount` as first param. Maybe you can rename this variable to that and avoid the underscore syntax/name shadowing.

https://firebase.google.com/docs/reference/admin/node/admin.credential#cert

Reply:
The logic here currently resolves the credentials from this script file's location, then passess it into `.cert` below.

`.cert`'s first param is called `serviceAccountPathOrObject`. Since we haven't actually loaded the credential file, in our usecase, it would basically need to be called `serviceAccountPath` to be accurate. Which is in essence what we are using with `credentialPath` (yet sticking to what we traditionally called this across all of our other scripts)

The `_` syntax + 'shadowing' here is basically a workaround to setting the default value on the parameter when not supplied, to simplify the typing, but for all intents and purposes, it should be considered the same as the param (which is why it was 'shadowed' in this way)

I believe the reasoning behind this (as I remember it) is that if we didn't do it this way, then we would either have to make `credentialPath` non-optional in the typing (which we don't really want to do), or we would need to specify the default value in 2 places (to satisfy both the 'normal' default, and then the default again in the empty object(`{}`) part of the props.

--

I don't think the `_` syntax + 'shadowing' is a big deal here, particularly how 'localised' it is within a helper function that isn't going to need to be changed much/ever. So i'll probably leave it as is.

Reply:
Sounds good.  It wasn't a change request or something that I strongly want to be changed. Keep it as it is 🎉